core/src/main/java/io/undertow/server/handlers/accesslog/DefaultAccessLogReceiver.java

```diff
@@ -153,75 +154,66 @@ public void logMessage(final String message) {
+                final String msg = pendingMessages.peek();
+                if (msg == null) {
+                    break;
                 }
+                writeMessage(msg);
+
+                // NOTE:this is very similar to remove(), but without screenNull
+                // at best, it will work like poll/remove, at worst, will do nothing
+                if (!pendingMessages.remove(msg)) {
+                    break;
+                }
+            }
+        }finally {
+            // flush what we might have
+            try {
+                //this can happen when log has been rotated and there were no write
+                final BufferedWriter bw = this.writer;
+                if(bw != null)
+                    bw.flush();
+            } catch (IOException e) {
+                UndertowLogger.ROOT_LOGGER.errorWritingAccessLog(e);
             }
-            // Check the state before resetting the state to 0 (not running) and checking if a writer needs to be closed:
-            // - If state != 3 here, another thread is executing this.
-            //   The other thread will visit here and will check if a writer needs to be closed.
-            //   We can leave state and skip closing a writer.
-            // - If state == 3 here, there is no another thread executing this.
-            //   So, update the state to 0 (not running) and check if a writer needs be closed.
-            if (stateUpdater.compareAndSet(this, 3, 0) && closed) {
-                // As close() can be invoked from another thread in parallel,
-                // it will dispatch a new thread to close writer if state == 0 (not running) at that moment.
-                // So, just in case, check the state again:
-                // - if state != 0, another thread has already dispatched from close() and it will visit here. So, closing writer can be skipped here.
-                // - if state == 0, writer can be closed here. Let's change state to 3 again in order to prevent close() from dispatching a new thread.
-                if (stateUpdater.compareAndSet(this, 0, 3)) {
-                    try {
-                        if(writer != null) {
-                            writer.flush();
-                            writer.close();
-                            writer = null;
-                        }
-                    } catch (IOException e) {
-                        UndertowLogger.ROOT_LOGGER.errorWritingAccessLog(e);
-                    } finally {
-                        // reset the state to 0 again finally
-                        stateUpdater.set(this, 0);
+            if(this.closed) {
+                if (!stateUpdater.compareAndSet(this, 2, 3)) {
+                    UndertowLogger.ROOT_LOGGER.accessLogWorkerFailureOnTransition();
+                }
+                return;
+            } else {
+                if (!pendingMessages.isEmpty() || forceLogRotation) {
```

Comment:
I have been thinking... we still need a state for being at the finally block, as it prevents us from having more than one thread scheduling a new execution of run at the same time

Reply:
how? if there is nothing in pendingMessages, it does not need scheduling, if someone writes, it will be able to schedule. In reality, pendingMessages are filled before state change, so its just a race between logMessage thread and run() to schedule new execution.